### DISCUSSION TITLE
TIFF to pdf conversion of 4.1GB tiff file gets failed with "UnableToExtendCache" error

### DISCUSSION THREAD

**Discussion by Sasidhar124** [ImageMagick/ImageMagick#5089] [Help]
### ImageMagick version

7.1.0.7 64 bit version

### Operating system

Windows

### Operating system, version and so on

Windows Server 2019 Datacenter, 8GB RAM, 64bit OS

### Description

Hi I am trying to convert a 4.19GB tiff file to pdf. The input file has 897 pages, with 300dpi, and 2550 x 4200 resolution. I have created a user environment variable "MAGICK_TEMPORARY_PATH" and assigned "C:\ImageMagickTemp" which magick.exe is using it for page conversions. The C drive has over 43GB free space. Our customer has reported this issue and we are also able to reproduce the issue. 

Is this issue caused due to the 4GB limitation of libtiff tool used by Imagemagick?

Please do let me know if I need to do anything wrong and help us solve the issue as early as possible.

### Steps to Reproduce

Below are the different arguments I have used for conversion with magick.exe and convert.exe: -

C:\Program Files (x86)\Kofax\KIC-ED\MC\bin\ImageMagick64>convert.exe -verbose C:\Large_Tiffs\4.19gb-897PAGES.tif -density 200 C:\Large_Tiffs\output\output419gb.pdf
convert.exe: UnableToExtendCache 'C:\Large_Tiffs\4.19gb-897PAGES.tif': No space left on device @ error/cache.c/OpenPixelCache/3902.
convert.exe: NoImagesDefined `C:\Large_Tiffs\output\output419gb.pdf' @ error/convert.c/ConvertImageCommand/3322.

C:\Program Files (x86)\Kofax\KIC-ED\MC\bin\ImageMagick64>convert.exe -verbose C:\Large_Tiffs\4.19gb-897PAGES.tif -quality 75 C:\Large_Tiffs\output\output419gb.pdf
convert.exe: UnableToExtendCache 'C:\Large_Tiffs\4.19gb-897PAGES.tif': No space left on device @ error/cache.c/OpenPixelCache/3902.
convert.exe: NoImagesDefined `C:\Large_Tiffs\output\output419gb.pdf' @ error/convert.c/ConvertImageCommand/3322.

C:\Program Files (x86)\Kofax\KIC-ED\MC\bin\ImageMagick64>convert.exe -verbose C:\Large_Tiffs\4.19gb-897PAGES.tif -resize 800*640! C:\Large_Tiffs\output\output419gb.pdf
convert.exe: UnableToExtendCache 'C:\Large_Tiffs\4.19gb-897PAGES.tif': No space left on device @ error/cache.c/OpenPixelCache/3902.
convert.exe: InvalidArgument '-resize': 800*640! @ error/convert.c/ConvertImageCommand/2652.

C:\Program Files (x86)\Kofax\KIC-ED\MC\bin\ImageMagick64>convert.exe -verbose C:\Large_Tiffs\4.19gb-897PAGES.tif -resize 800*640\! C:\Large_Tiffs\output\output419gb.pdf
convert.exe: UnableToExtendCache 'C:\Large_Tiffs\4.19gb-897PAGES.tif': No space left on device @ error/cache.c/OpenPixelCache/3902.
convert.exe: InvalidArgument '-resize': 800*640\! @ error/convert.c/ConvertImageCommand/2652.

C:\Program Files (x86)\Kofax\KIC-ED\MC\bin\ImageMagick64>magick.exe -verbose C:\Large_Tiffs\4.19gb-897PAGES.tif -resize 800*640\! C:\Large_Tiffs\output\output419gb.pdf
magick.exe: UnableToExtendCache 'C:\Large_Tiffs\4.19gb-897PAGES.tif': No space left on device @ error/cache.c/OpenPixelCache/3902.

C:\Program Files (x86)\Kofax\KIC-ED\MC\bin\ImageMagick64>magick.exe -verbose C:\Large_Tiffs\4.19gb-897PAGES.tif -limit memory 500mb -limit map 500mb C:\Large_Tiffs\output\output419gb.pdf
magick.exe: UnableToExtendCache 'C:\Large_Tiffs\4.19gb-897PAGES.tif': No space left on device @ error/cache.c/OpenPixelCache/3902.

### Images

![errors](https://user-images.githubusercontent.com/104485348/165464096-8381bb0c-3e05-4533-99ca-7b6c04a77db8.png)

**Comment by dlemstra** [comment]
What is the output of `convert -version` and `convert C:\Large_Tiffs\4.19gb-897PAGES.tif[0] info:`

**Comment by Sasidhar124** [comment]
Hi Dirk,
Thank you for the quick response. Please find below the response for each command

convert -version:
Version: ImageMagick 7.1.0-7 Q8-HDRI x64 2021-09-13 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2019 (192829914)


convert -density 300 C:\Large_Tiffs\4.19gb-897PAGES.tif[0] info:
C:\Large_Tiffs\4.19gb-897PAGES.tif[0] TIFF 2550x4200 2550x4200+0+0 8-bit TrueColor sRGB 3.996GiB 3.672u 0:06.585

Thanks and regards,
Sasidhar


Venkat Kambhampati
Principal Software Engineer

[cid:Kofax_ab6d7362-d88f-454c-a5e0-9fbdacd27b9c.jpg]

Kofax India Private Limited
V IT Park, Capella Building, Floor 6, Plot No. 17, Software Units Layout
Hyderabad, Madhapur 500081

Tel: +91 40 45459949
Fax: +91 40 45459922
***@***.***

 [cid:WorkLikeTomorrow_5b748fe4-9e86-42f8-b5af-b79a2d6ee254.jpg]

.

______________________________________________________________________________________________________________________________________________________________

This communication is only for the use of the intended recipient. It may contain confidential or proprietary information. If you are not the intended recipient or have received this communication in error, please notify the sender via phone and destroy this communication immediately.


From: Dirk Lemstra ***@***.***>
Sent: 27 April 2022 13:47
To: ImageMagick/ImageMagick ***@***.***>
Cc: Venkat Kambhampati ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [ImageMagick/ImageMagick] TIFF to pdf conversion of 4.1GB tiff file gets failed with "UnableToExtendCache" error (Issue #5088)


What is the output of convert -version and convert -density 300 C:\Large_Tiffs\4.19gb-897PAGES.tif[0] info:

-

**Comment by dlemstra** [comment]
Because you are using a HDRI version of ImageMagick you need 32 bits per channel. Assuming your input image has three channels you will need `32*3*2550*4200 = 1028160000` bits per image. And because you are reading all 897 pages ImageMagick will need at least  `32*3*2550*4200*897  = 922259520000` bits (and that is around 115 GB) of memory or hard disk space. 

If you would switch to just the Q8 version you would "only" require `8*3*2550*4200*897 = 230564880000` bits (around 28 GB).

**Reply by Sasidhar124** [reply]
Hello dlemstra, 
Thank you for the suggestion, Now I am trying to build the  Q8 bit version ImageMagick source code v7.1.0.31 using VS 2019. But I am getting build errors for 162 modules. All modules fail with the below error from winbase.h. Am I missing anything? Thanks in advance..

Error	C2220	the following warning is treated as an error	UTIL_identify	C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\um\winbase.h	9531	

Warning	C5105	macro expansion producing 'defined' has undefined behavior	IM_MOD_ttf	C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\um\winbase.h	9531	

Warning	C4244	'=': conversion from 'const _Ty' to 'float', possible loss of data	CORE_jpeg-xl	C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\include\xutility	4595	

Similar errors are repeated for 162 modules..

As per readme.md file from imagemagick github, I have installed VS 2019 profession with following modules in place. even same errors are seen with 2017 as well.
- Desktop development with C++ (workload)
- Visual C++ ATL for x86 and x64
- Visual C++ MFC for x86 and x64

Thanks you,
Sasidhar

**Reply by Sasidhar124** [reply]
HI
At the time of opening VisualDynamicMT.sln for the first time I have set the target platform to 10.0.19041.0. Now I am getting compilation error like undefined variable errors. I am missing something?

Thank you,
Sasidhar

**Reply by dlemstra** [reply]
You will probably need to install a newer version of the Windows SDK to make it build correctly. There is a specific version that causes these kinds of errors.

**Comment by Sasidhar124** [comment]
Hi  ,
 I have installed latest update of Visual studio build tools  and have set the target platform toolset to v142 in VisualDynamicMT.sln
But I am getting the same errors. Can anyone confirm whether it is working fine for anyone in ImageMagick on windows 10 OS x64.

Thanks and regards,
Sasidhar



Venkat Kambhampati
Principal Software Engineer

[cid:Kofax_ab6d7362-d88f-454c-a5e0-9fbdacd27b9c.jpg]

Kofax India Private Limited
V IT Park, Capella Building, Floor 6, Plot No. 17, Software Units Layout
Hyderabad, Madhapur 500081

Tel: +91 40 45459949
Fax: +91 40 45459922
***@***.***

 [cid:WorkLikeTomorrow_5b748fe4-9e86-42f8-b5af-b79a2d6ee254.jpg]

.

______________________________________________________________________________________________________________________________________________________________

This communication is only for the use of the intended recipient. It may contain confidential or proprietary information. If you are not the intended recipient or have received this communication in error, please notify the sender via phone and destroy this communication immediately.


From: Dirk Lemstra ***@***.***>
Sent: 29 April 2022 15:53
To: ImageMagick/ImageMagick ***@***.***>
Cc: Venkat Kambhampati ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [ImageMagick/ImageMagick] TIFF to pdf conversion of 4.1GB tiff file gets failed with "UnableToExtendCache" error (Discussion #5089)


You will probably need to install a newer version of the Windows SDK to make it build correctly. There is a specific version that causes these kinds of errors.

-

**Comment by dlemstra** [comment]
As I said before you will probably need to upgrade the version of the Windows SDK that is being used.

**Comment by Sasidhar124** [comment]
HI Dirk,
I think it would be of great help,  if IMagemagick could specify the compatible windows SDK version for building the source code. For Windows 10 the latest version of SDK is 10.0.19041.0 which I am already using it for the build and still I am getting the build errors.

Thanks and regards,
Sasidhar


Venkat Kambhampati
Principal Software Engineer

[cid:Kofax_ab6d7362-d88f-454c-a5e0-9fbdacd27b9c.jpg]

Kofax India Private Limited
V IT Park, Capella Building, Floor 6, Plot No. 17, Software Units Layout
Hyderabad, Madhapur 500081

Tel: +91 40 45459949
Fax: +91 40 45459922
***@***.***

 [cid:WorkLikeTomorrow_5b748fe4-9e86-42f8-b5af-b79a2d6ee254.jpg]

.

______________________________________________________________________________________________________________________________________________________________

This communication is only for the use of the intended recipient. It may contain confidential or proprietary information. If you are not the intended recipient or have received this communication in error, please notify the sender via phone and destroy this communication immediately.


From: Dirk Lemstra ***@***.***>
Sent: 29 April 2022 19:31
To: ImageMagick/ImageMagick ***@***.***>
Cc: Venkat Kambhampati ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [ImageMagick/ImageMagick] TIFF to pdf conversion of 4.1GB tiff file gets failed with "UnableToExtendCache" error (Discussion #5089)


As I said before you will probably need to upgrade the version of the Windows SDK that is being used.

-

**Comment by dlemstra** [comment]
We have a daily build that checks if we can compile ImageMagick on Windows without any errors and we can so this must be something in your local setup.

**Comment by Sasidhar124** [comment]
HI Dirk,
I am able to successfully build using windows SDK version 10.0.22000.0. I will now test the binaries and see.

Thank you so much for your assistance.

Sasidhar


Venkat Kambhampati
Principal Software Engineer

[cid:Kofax_ab6d7362-d88f-454c-a5e0-9fbdacd27b9c.jpg]

Kofax India Private Limited
V IT Park, Capella Building, Floor 6, Plot No. 17, Software Units Layout
Hyderabad, Madhapur 500081

Tel: +91 40 45459949
Fax: +91 40 45459922
***@***.***

 [cid:WorkLikeTomorrow_5b748fe4-9e86-42f8-b5af-b79a2d6ee254.jpg]

.

______________________________________________________________________________________________________________________________________________________________

This communication is only for the use of the intended recipient. It may contain confidential or proprietary information. If you are not the intended recipient or have received this communication in error, please notify the sender via phone and destroy this communication immediately.


From: Dirk Lemstra ***@***.***>
Sent: 30 April 2022 14:52
To: ImageMagick/ImageMagick ***@***.***>
Cc: Venkat Kambhampati ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [ImageMagick/ImageMagick] TIFF to pdf conversion of 4.1GB tiff file gets failed with "UnableToExtendCache" error (Discussion #5089)


We have a daily build that checks if we can compile ImageMagick on Windows without any errors and we can so this must be something in your local setup.

-
